rust/ql/lib/codeql/rust/dataflow/internal/TaintTrackingImpl.qll

```diff
@@ -46,6 +46,8 @@ module RustTaintTracking implements InputSig<Location, RustDataFlow> {
         RustDataFlow::readStep(pred, cs, succ) and
         cs.getContent() instanceof ArrayElementContent
       )
+      or
+      pred.asExpr() = succ.asExpr().(RefExprCfgNode).getExpr()
```

Comment:
Yes, that is right. I added the taint flow to support this line in the SQL injection test:

```rust
let unsafe_query_1 = String::from("SELECT * FROM people WHERE firstname='") + &remote_string + "'";
```

Here `remote_string` is tainted, and the extra taint step makes `unsafe_query_1` tainted at well. One could argue that the reference itself isn't really tainted, but on the other hand the only thing it can be used for is access tainted data and it seemed like a simple way to unlock some additional flow. Alternatively, we could also extend the handling of `+` to read `ReferenceContent` as well?

Reply:
My intuition is that having `+` read the `ReferenceContent` is more accurate but ... I'm worried this will be a can of worms if we got this way.  So I guess we should probably leave it the way it is.

@hvitved do you have an opinion on this?

Reply:
Modelling store steps as also taint steps has proven bad in the past, so I think it would be better to provide a taint flow summary for `+` which pops `ReferenceContent`.

Reply:
What would be the best way to do that for a built-in operator?

Reply:
Ideally, the extractor should supply the information (`a + b` is [syntactic sugar for](https://doc.rust-lang.org/rust-by-example/trait/ops.html) `a.add(b)`), but until that happens, we could add synthetic `DataFlowCall`s.

Reply:
In any case, I think we should do that in a separate PR.

Reply:
I still think we should revert this.

Reply:
> Modelling store steps as also taint steps has proven bad in the past

Re. this, we also do that right now for arrays (which was inspired by Ruby). Do we want to remove that as well (later)?

Reply:
Hopefully we only add taint steps for _reads_ out of arrays, and nor for stores into arrays?

Reply:
Yes, that's right. Got it, taint steps for read steps are fine, but taint steps for store steps are not.

Reply:
> I still think we should revert this.

Done 👍